In [7]:
''''
NAME:YANG ZITAO
student NO.:1155098948
''''
#For question (a)
#black_scholes model
from scipy import stats
import math

def black_scholes (cp, s, k, t, v, rf):
        """ Price an option using the Black-Scholes model.
        s: initial stock price
        k: strike price
        t: expiration time
        v: volatility
        rf: risk-free rate
        cp: +1/-1 for call/put
        """

        d1 = (math.log(s/k)+(rf-div+0.5*math.pow(v,2))*t)/(v*math.sqrt(t))
        d2 = d1 - v*math.sqrt(t)

        optprice = (cp*s*stats.norm.cdf(cp*d1)) - (cp*k*math.exp(-rf*t)*stats.norm.cdf(cp*d2))
        return optprice

S = 40
K = 42
r = 0.1
v = 0.2
T = 0.5
cp = 1
price = black_scholes(cp,S,K,T,v,r)
print('Using the Black-Scholes pricing, the option price is ' + str(price))

Using the Black-Scholes pricing, the option price is 2.27778032946


In [4]:
#For question (b)
#binomial tree model
import math

class Bnode(object):
    def __init__(self,price,e_value,style):
        self.price = price
        self.e_value = e_value
        self.c_value = 0
        self.style = style

    def setcvalue(self,c_value):
        #continue value
        #also the value of the options in European option
        self.c_value = c_value
        if self.style == 'A':
            #American
            self.o_value = max(self.c_value, self.e_value)
        else:
            #European
            self.o_value = c_value
    def __repr__(self):
        return '(' + str(self.price) +','+ str(self.e_value) + ',' + str(self.c_value) + ',' + str(self.o_value) + ')'

def binTree(S,K,r,v,T,n,otype,style='E'):
    ''''Price an option using the Black-Scholes model
    S is the initial price
    K is the strike price at the maturity
    r is the risk-free interest rate
    v is the volatility
    T is the length of period
    n is the number of steps
    otype is the call/put. input 'call'means call option, others means put
    style='A' means American options while 'E' means European option
    ''''
    delta_t = T/n
    
    a = math.exp(r*delta_t)
    u = math.exp(r*delta_t + v*math.sqrt(delta_t))
    d = math.exp(r*delta_t - v*math.sqrt(delta_t))
    
    p = (a-d)/(u-d)
    q = 1-p

    root = Bnode(S,0,style)
    price_list = [[root]]
    
    for l in range(1,n+1):
        new_layer = []
        for i in range(l+1):
            price = S*(u**(l-i))*(d**i)
            if otype == 'call':
                node = Bnode(price,max(price-K,0),style)#for call option
            else:
                node = Bnode(price,max(K-price,0),style)#for put option
            new_layer.append(node)
        price_list.append(new_layer)

    #the value of end node is the same as the difference with strike price
    for each in price_list[-1]:
        each.setcvalue(each.e_value)
    
    for index, line in enumerate(price_list[-2::-1]):
        children = price_list[-index-1]
        for i,each in enumerate(line):
            each.setcvalue((p*children[i].o_value + q*children[i+1].o_value)/a)
    print('Using the binomial tree pricing, the option price is '+ str(price_list[0][0].o_value))


S = 40
K = 42
r = 0.1
v = 0.2
T = 0.5
n = 1
otype = 'call'
binTree(S,K,r,v,T,n,otype,style='E')

Using the binomial tree pricing, the option price is 2.846197486606504


SUMMARY:
Using the Black-Scholes pricing, the option price is 2.27778032946,
Using the binomial tree pricing, the option price is 2.84619748661.
the price generated from binomial tree model is larger than that from Black-Scholes model.